[postfilter-impl-3] Register and enable the defaultpreemption plugin

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind feature
/sig scheduling

**What this PR does / why we need it**:

This PR registers and enable defaultpreemption plugin. At the same time, legacy hard-coded preemption call in scheduler.go is removed.

**Which issue(s) this PR fixes**:

Part of #91038.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
A defaultpreemption plugin is registered and enabled in scheduler which replaces the legacy hard-coded Pod preemption logic.
```

Huang-Wei commented:
/test all

Huang-Wei commented:
/test all

Huang-Wei commented:
/test all

Huang-Wei commented:
/test all

Huang-Wei commented:
@ahg-g I resolved some of your comments in #92108 to this PR.

Huang-Wei commented:
/retest

Huang-Wei commented:
@ahg-g this pr is ready for review.

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -525,6 +525,13 @@ func (f *framework) runFilterPlugin(ctx context.Context, pl FilterPlugin, state
 // RunPostFilterPlugins runs the set of configured PostFilter plugins until the first
 // Success or Error is met, otherwise continues to execute all plugins.
 func (f *framework) RunPostFilterPlugins(ctx context.Context, state *CycleState, pod *v1.Pod, filteredNodeStatusMap NodeToStatusMap) (*PostFilterResult, *Status) {
+	preemptionStartTime := time.Now()
+	defer func() {
+		metrics.PreemptionAttempts.Inc()
+		metrics.SchedulingAlgorithmPreemptionEvaluationDuration.Observe(metrics.SinceInSeconds(preemptionStartTime))
+		metrics.DeprecatedSchedulingDuration.WithLabelValues(metrics.PreemptionEvaluation).Observe(metrics.SinceInSeconds(preemptionStartTime))
+	}()
+
```
this should be inside the preemption plugin. We need a separate PostFilter extension point metric like the others.

Huang-Wei replied:
Do you mean the metric defined in `runPostFilterPlugin()`?

ahg-g replied:
No, I am talking about `FrameworkExtensionPointDuration`, see RunScorePlugins for example.

```PreemptionEvaluation``` should only be measuring the default preemption logic, and so we should move it inside the plugin. We can later deprecate this metric since this overhead will be covered by the metric set in `runPostFilterPlugin`.

Huang-Wei replied:
Done. PTAL.

cc/ @alculquicondor

ahg-g replied:
Sounds good. PreemptionEvaluation latency is already covered by the extension point latency metric, and so PreemptionEvaluation can be deprecated later. But we don't have a PreemptionAttempts equivalent counter, should we have one for extension point attempts?

alculquicondor replied:
+1 to deprecating Preemption latency in the future.

Huang-Wei commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -44,7 +48,10 @@ func (pl *DefaultPreemption) Name() string {
-	pl := DefaultPreemption{fh}
+	pl := DefaultPreemption{fh: fh}
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.PodDisruptionBudget) {
+		pl.pdbLister = fh.SharedInformerFactory().Policy().V1beta1().PodDisruptionBudgets().Lister()
```
@ahg-g to avoid temporary signature change, the pdbLister is defined here but not used in Preempt(). I will use the pdbLister in the next followup - where Preempt() gets moved to defaultpreemption package.

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

alculquicondor commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -525,7 +525,12 @@ func (f *frameworkImpl) runFilterPlugin(ctx context.Context, pl framework.Filter
 // RunPostFilterPlugins runs the set of configured PostFilter plugins until the first
 // Success or Error is met, otherwise continues to execute all plugins.
-func (f *frameworkImpl) RunPostFilterPlugins(ctx context.Context, state *framework.CycleState, pod *v1.Pod, filteredNodeStatusMap framework.NodeToStatusMap) (*framework.PostFilterResult, *framework.Status) {
+func (f *frameworkImpl) RunPostFilterPlugins(ctx context.Context, state *framework.CycleState, pod *v1.Pod, filteredNodeStatusMap framework.NodeToStatusMap) (r *framework.PostFilterResult, s *framework.Status) {
```
we are already calling the variable `status` in all other functions

Huang-Wei replied:
which variable? the returned status?

alculquicondor replied:
correct

Huang-Wei replied:
Ok I will update along with the squash

Huang-Wei replied:
done.

Huang-Wei commented:
@ahg-g please take an another look.

ahg-g commented:
Looks good to me, please squashed.

Huang-Wei committed:
Register and enable defaultpreemption plugin

- Enable defaultpreemption as a PostFilter plugin
- Remote legacy hard-coded preemption logic

Huang-Wei commented:
@ahg-g squashed.

ahg-g commented:
/lgtm
